incubator/hnc/pkg/reconcilers/hnc_config.go

```diff
@@ -136,28 +138,52 @@ func (r *ConfigReconciler) createObjectReconcilers(inst *api.HNCConfiguration) {
 	// from the forest for the object reconciliation, after we create ObjectReconcilers but before we write the
 	// ObjectReconcilers to the types list. As a result, objects of the newly added types might not be propagated correctly
 	// using the latest forest structure.
+	//
+	// For syncing an object reconciler, the mutex is guarding the read access to the `namespaces` field in the forest. The
+	// `namespaces` field is a shared resource between various reconcilers.
 	r.Forest.Lock()
 	defer r.Forest.Unlock()
 
 	for _, t := range inst.Spec.Types {
 		gvk := schema.FromAPIVersionAndKind(t.APIVersion, t.Kind)
-		if r.Forest.HasTypeSyncer(gvk) {
-			continue
+		mode := r.getMode(t)
+		if ts := r.Forest.GetTypeSyncer(gvk); ts != nil {
+			r.syncObjectReconciler(ctx, mode, ts)
+		} else {
+			r.createObjectReconciler(gvk, mode, inst)
+		}
+	}
+}
+
+// syncObjectReconciler syncs the mode of an ObjectReconciler. It also updates corresponding objects, if needed.
+func (r *ConfigReconciler) syncObjectReconciler(ctx context.Context, mode api.SynchronizationMode, ts forest.TypeSyncer) {
+	r.Log.Info("Syncing an object reconciler.", "gvk", ts.GetGVK())
+
+	oldMode := ts.GetMode()
+	if mode != oldMode {
+		r.Log.Info("Changing Mode of the object reconciler", "old Mode", oldMode, "new Mode", mode)
+		ts.SetMode(mode)
+		// If the new mode is "propagate" or "remove", we need to update objects in the cluster
+		// (e.g., propagate or remove existing objects).
+		if mode == api.Propagate || mode == api.Remove {
+			err := ts.SyncCluster(ctx, r.Log)
```

Comment:
Haven't looked into the tests, but all other lgtm. WDYT about changing `SyncCluster` to `EnqueueAllObjects`?

Reply:
Good idea. `EnqueueAllObjects` is more precise. Feel free to see updated code in `object.go`.